### PR TITLE
Gemified repository; Included models for QDM 5.3

### PR DESCRIPTION
**Pull request by holmesie**
This PR includes an updated script to generate Ruby (mongoid) and JavaScript (mongoose) models from a modelinfo file (describes the QDM datatypes and their attributes). This PR also includes models generated for QDM 5.3. The repository is also now a Gem, so it can be made a dependency by Ruby projects.

---

Pull requests into cqm-models require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] Internal ticket for this PR: https://jira.mitre.org/browse/TACO-34, https://jira.mitre.org/browse/BONNIE-1265
- [x] Internal ticket links to this PR
- [x] Code diff has been done and been reviewed
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass

**Reviewer 1:**

Name: @okeefm
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/.bundle/
+/.yardoc
+/Gemfile.lock
```
Comment (okeefm) [inline .gitignore]: I'm not sure if it's best practice to gitignore `Gemfile.lock`. Is that specific to a Rubygem?

File: spec/cqm/models_spec.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+require "spec_helper"
+
+RSpec.describe Cqm::Models do
+  it "has a version number" do
+    expect(Cqm::Models::VERSION).not_to be nil
+  end
+
+  it "does something useful" do
+    expect(false).to eq(true)
```
Comment (okeefm) [inline spec/cqm/models_spec.rb]: This seems...wrong, somehow.

File: .gitignore (hunk 1 above)
Comment (holmesie) [reply .gitignore]: Yes, you should never include the lockfile when writing a Gem. The .gemspec takes care of dependencies and their versions.

File: spec/cqm/models_spec.rb (hunk 2 above)
Comment (holmesie) [reply spec/cqm/models_spec.rb]: Yes, indeed it is. I'm going to add tests for each version of QDM that is included (so a test for each included modelinfo file), so basically some sanity checks to make sure the generator is handling each as expected. I'll ping you when finished.

File: templates/mongoose_template.js.erb [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
-    <% end %>
-  <% end %>
+var <%= datatype -%>Schema = new Schema({
+  <%- for attribute in attributes -%>
+  <%- if attributes.last[:name] == attribute[:name] -%>
+  <%= attribute[:name] %>: <%= MODELINFO_JS_LOOKUP[attribute[:type]] -%>
+  <%- else -%>
+  <%= attribute[:name] %>: <%= MODELINFO_JS_LOOKUP[attribute[:type]] -%>,
+  <%- end -%>
+  <%- end %>
```
Comment (ljades) [inline templates/mongoose_template.js.erb]: Any reason why all the indentation was removed in this block?

Comment (holmesie) [reply templates/mongoose_template.js.erb]: Because the only lines that actually print anything are the `<%=` ones, and we only want a two space indentation, rather than a six.

File: app/models/qdm/patient.rb [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    history_elements.where(oid: oid) || []
+  end
+
+  # Returns an array of history elements that exist on this patient. Optionally
+  # takes a type, which returns all history elements of that QDM category.
+  # Example: patient.get_history_elements(type='encounters') will return all
+  # Encounter QDM data types active on the patient.
+  def get_history_elements(type=nil)
+    if type
+      history_elements.where(category: type)
```
Comment (okeefm) [inline app/models/qdm/patient.rb]: rather than do a `|| []` in each getter, does it make more sense to guarantee that `get_history_elements` will always return an `Array`? That way all the `history_element` getters will be consistent too.

File: app/models/models.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
+module Qdm
```
Comment (okeefm) [inline app/models/models.rb]: Quick question: do we want to capitalize all the letters in `QDM` when used as a module name? I think that's what a lot of libraries do with acronyms.

File: app/models/qdm/patient.rb (hunk 4 above)
Comment (holmesie) [reply app/models/qdm/patient.rb]: That does make good sense, I'll put in that change.

File: app/models/models.rb (hunk 5 above)
Comment (holmesie) [reply app/models/models.rb]: Yeah, I'll take a look at this.

File: app/assets/javascripts/Patient.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+var Code = Schema.Types.Code;
+var Interval = Schema.Types.Interval;
+var Quantity = Schema.Types.Quantity;
+
+var PatientSchema = new Schema({
+  birth_datetime: DateTime,
+  qdm_version: { type: String, default: "5.3" },
+  ethnicity: Code,
+  race: Code,
+  sex: Code,
```
Comment (jbradl11) [inline app/assets/javascripts/Patient.js]: Does "payer" and "expired" need to be listed here?

File: app/assets/javascripts/Patient.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+PatientSchema.methods.care_experiences = function care_experiences(params, callback) {
+  return this.get_history_elements('care_experience');
+}
+
+PatientSchema.methods.preferences = function preferences(params, callback) {
+  return this.get_history_elements('preference');
+}
+
+PatientSchema.methods.provider_characteristics = function provider_characteristics(params, callback) {
```
Comment (jbradl11) [inline app/assets/javascripts/Patient.js]: Do we need to add a function for 'patient_characteristic' which is the category for the 'payer' model?

Comment (holmesie) [reply app/assets/javascripts/Patient.js]: All of the patient characteristics (minus expired, which is a condition) are fields on the patient itself, rather than entries in the `history_elements`.

That being said, those characteristics _ARE_ still models coming from the modelinfo file, which theoretically could at some point be entries in `history_elements`, so I think it does make sense to have a helper method for those types specifically. I will add that in!

File: app/models/qdm/patient.rb [hunk 8]
```diff
@@ -0,0 +1,143 @@
+class QDM::Patient
+  include Mongoid::Document
+  field :birth_datetime, type: DateTime
+  field :qdm_version, type: String, default: '5.3'
+  field :ethnicity, type: QDM::Code
+  field :race, type: QDM::Code
+  field :sex, type: QDM::Code
```
Comment (jbradl11) [inline app/models/qdm/patient.rb]: Same questions that I asked above on the Javascript model. Do we need to add 'payer' and add an additional function for the 'patient_characteristic' category?

File: app/assets/javascripts/Patient.js (hunk 6 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: Like `ethnicity`, `race`, and `sex` - `payer` and `expired` are also actual QDM datatypes, which means there are models for them.

I think we decided that `ethnicity`, `race`, and `sex` should be on the patient itself as well, because that is similar to how HDS did things.

So I guess my question is, do we want to do this for _ALL_ of the QDM patient characteristic types? I'm not opposed to that, I just want to clarify that there is duplication by doing this.

An alternative might be using the patient characteristic datatypes themselves (so initialize them using the QDM models, and put them inside the `history_elements`), rather than sticking that information on the patient model itself, and instead have helper methods (like `patient.procedures`, `patient.encounters`, etc) that you can call on the patient (i.e. `patient.race`), which looks into `history_elements`, grabs any existing data criteria for race, and returns what it finds.

File: app/models/qdm/patient.rb (hunk 8 above)
Comment (holmesie) [reply app/models/qdm/patient.rb]: See previous reply ☝️

File: app/assets/javascripts/Patient.js (hunk 6 above)
Comment (edeyoung) [reply app/assets/javascripts/Patient.js]: i like the alternative

Comment (edeyoung) [reply app/assets/javascripts/Patient.js]: if we do the alternative, does it make sense to rename "history elements" to something else? don't have many ideas here - maybe "attributes"?

File: app/assets/javascripts/Patient.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+// all Encounter QDM data types active on the patient.
+PatientSchema.methods.get_history_elements = function get_history_elements(category, callback) {
+  return this.history_elements.filter(element => element.category === category);
+}
+
+PatientSchema.methods.adverse_events = function adverse_events(params, callback) {
+  return this.get_history_elements('adverse_event');
+}
+
+PatientSchema.methods.assessments = function assessments(params, callback) {
```
Comment (edeyoung) [inline app/assets/javascripts/Patient.js]: this method seems a bit like the HDS way which has gotten us into trouble. I think it could be different here because it directly maps to the QDM model and isn't related to parsing things, but it makes me nervous based on the issues we've had.

There should be a way to access by OID. If that were present, then both the oid method and this would be fine.

File: app/assets/javascripts/Patient.js [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+var PatientSchema = new Schema({
+  birth_datetime: DateTime,
+  qdm_version: { type: String, default: "5.3" },
+  ethnicity: Code,
+  race: Code,
+  sex: Code,
+  // These are the "data criteria", or QDM datatype elements that exist on a
+  // patient.
+  history_elements: []
```
Comment (edeyoung) [inline app/assets/javascripts/Patient.js]: It looks like everything is being stored in an array. Will this cause access efficiency issues over time? Does it make sense to store these in a hash based on the data type or oid of the data type?

File: app/assets/javascripts/Patient.js (hunk 6 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: Right, we had a long discussion on slack about how to name it, and I don't think any of us were too attached to any of the options. I do agree it should be changed with the understanding that we are going with the "alternative" above.

Attributes sounds nice, but unfortunately if we used that here it would be overloading the term, as the QDM spec uses it to describe fields on QDM datatypes.

We also considered: `entries`, `data_criteria`, `data_criteria_entries`, `patient_history_elements`, `patient_history_entries`

File: app/assets/javascripts/CareGoal.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
+var Schema = mongoose.Schema;
+var Code = Schema.Types.Code;
+var Interval = Schema.Types.Interval;
+var Quantity = Schema.Types.Quantity;
+
+var CareGoalSchema = new Schema({
+  relevant_period: Interval,
+  related_to: Array,
+  target_outcome: {},
+  hqmf_oid: { type: String, default: "2.16.840.1.113883.10.20.28.3.7" },
```
Comment (edeyoung) [inline app/assets/javascripts/CareGoal.js]: happy to see this!

File: app/assets/javascripts/Patient.js (hunk 7 above)
Comment (edeyoung) [reply app/assets/javascripts/Patient.js]: i'm going to forward an email thread to you about patient characteristics. it sounds like measure developers can (and are interested in) adding random things in as patient characteristics. I don't think it makes sense to add here, but I'll send to you so you can take a look at it.

File: app/assets/javascripts/MedicationAdministered.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  author_datetime: DateTime,
+  relevant_period: Interval,
+  dosage: Quantity,
+  supply: Quantity,
+  frequency: Code,
+  route: Code,
+  reason: Code,
+  negation_rationale: Code,
+  hqmf_oid: { type: String, default: "2.16.840.1.113883.10.20.28.3.45" },
+  category: { type: String, default: "medication" },
```
Comment (edeyoung) [inline app/assets/javascripts/MedicationAdministered.js]: does it make sense to also have a "type"? What I mean by that in this case is "administered". It could be "ordered", "performed", etc.

File: app/assets/javascripts/Patient.js (hunk 9 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: These methods are just helper methods to grab data criteria on the patient by QDM category (Cypress needs this). This is different than the way things are with HDS at the moment, where the data_criteria are stored in collections dictated by their types (rather than in a single collection, like it is with this approach). This would make things easier for Bonnie (we wouldn't have to do things like this anymore: https://github.com/projecttacoma/cql_qdm_patientapi/blob/master/app/assets/javascripts/cqlpatient.js.coffee#L156).

There is a method above to grab by HQMF OID, is that what you mean?

Let me know if I misunderstood your question.

File: app/assets/javascripts/Patient.js [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+// match the given HQMF data criteria OID.
+PatientSchema.methods.get_by_hqmf_oid = function get_by_hqmf_oid(hqmf_oid, callback) {
+  return this.history_elements.filter(element => element.hqmf_oid === hqmf_oid);
+}
+
+// Returns an array of history elements that exist on this patient. Optionally
+// takes a category, which returns all history elements of that QDM category.
+// Example: patient.get_history_elements(category = 'encounters') will return
+// all Encounter QDM data types active on the patient.
+PatientSchema.methods.get_history_elements = function get_history_elements(category, callback) {
```
Comment (edeyoung) [inline app/assets/javascripts/Patient.js]: would it make sense to enable querying by category and type (ordered/performed/etc.?)

so i could explicitly search for "medication performed" or "medication ordered".

as i'm writing this, i'm thinking yes, especially as these are different models and not being able to do a direct query for them would mean I'd have to do some additional processing to figure out exactly what I'm looking at.

File: app/assets/javascripts/Patient.js (hunk 9 above)
Comment (edeyoung) [reply app/assets/javascripts/Patient.js]: totally missed the hqmf oid query possibility! thanks for pointing that out!

File: app/assets/javascripts/Patient.js [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+PatientSchema.methods.adverse_events = function adverse_events(params, callback) {
+  return this.get_history_elements('adverse_event');
+}
+
+PatientSchema.methods.assessments = function assessments(params, callback) {
+  return this.get_history_elements('assessment');
+}
+
+PatientSchema.methods.allergies = function allergies(params, callback) {
```
Comment (edeyoung) [inline app/assets/javascripts/Patient.js]: What is the maintenance plan for these fields as QDM evolves? Manual maintenance?

File: app/assets/javascripts/Patient.js [hunk 15]
```diff
@@ -0,0 +1,143 @@
+var Schema = mongoose.Schema;
+var Code = Schema.Types.Code;
+var Interval = Schema.Types.Interval;
+var Quantity = Schema.Types.Quantity;
+
+var PatientSchema = new Schema({
+  birth_datetime: DateTime,
```
Comment (edeyoung) [inline app/assets/javascripts/Patient.js]: Is this missing the name fields?

File: app/assets/javascripts/Patient.js (hunk 9 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: These are just helper methods to grab the criteria out of `history_elements` by QDM category (Cypress needs this). This differs from the way HDS does it currently (criteria are stored on the record by category: https://github.com/projectcypress/health-data-standards/blob/master/lib/health-data-standards/models/record.rb#L31). With this approach we are putting them all in one, but still allowing you to look them up by category. This will make things easier for Bonnie, as we won't need to do this anymore: https://github.com/projecttacoma/cql_qdm_patientapi/blob/master/app/assets/javascripts/cqlpatient.js.coffee#L156.

There is a `find_by_hqmf_oid` method above, is that what you mean?

Let me know if I misunderstood your question.

File: app/assets/javascripts/Patient.js (hunk 10 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: I need to double check, but I believe this is just how mongoose does `embeds_many`.

File: app/assets/javascripts/MedicationAdministered.js (hunk 12 above)
Comment (holmesie) [reply app/assets/javascripts/MedicationAdministered.js]: Thats what `category` basically is, it's how the QDM documentation refers to it anyways. Using `category` also saves us from confusion between using `type` as the literal mongo field type.

File: app/assets/javascripts/Patient.js (hunk 9 above)
Comment (edeyoung) [reply app/assets/javascripts/Patient.js]: this is like deja vu... yes, i see the find_by_hqmf_oid now :) saw it in the previous post although i guess my post didn't make that clear.

File: app/assets/javascripts/MedicationAdministered.js (hunk 12 above)
Comment (holmesie) [reply app/assets/javascripts/MedicationAdministered.js]: Yeah, that makes sense. I can't see any use for that off the top of my head, but it's easy enough to do.

File: app/assets/javascripts/Patient.js (hunk 13 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: Makes sense, thinking about the above comment (adding "types").

File: app/assets/javascripts/Patient.js (hunk 14 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: You mean methods? If QDM ever introduces new `categories`, then yeah.

File: app/assets/javascripts/Patient.js (hunk 15 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: Good catch!

File: app/assets/javascripts/Patient.js (hunk 7 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: There is a generic `patient_characteristic` model in the spec, with just one attribute (`author_datetime`). We could add another field to it (maybe `data`?) that could hold anything (Code, Quantity, Interval, etc.), and any time measure developers want to add whatever, they could use it.

Thoughts?

File: app/assets/javascripts/Patient.js (hunk 9 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: Ahh! Sorry, Github for some reason was not showing my original response (or your reply). I thought I forgot to submit the original post (and proceeded to re-write it)!

Comment (edeyoung) [reply app/assets/javascripts/Patient.js]: ha! :)

File: app/assets/javascripts/Patient.js (hunk 15 above)
Comment (edeyoung) [reply app/assets/javascripts/Patient.js]: what about expected values? and cached results?

Comment (edeyoung) [reply app/assets/javascripts/Patient.js]: should we create a data structure (where?) for result types? hash keyed by population, and then keyed by IPP/DENOM/MSRPOPLEX/etc.?

Comment (holmesie) [reply app/assets/javascripts/Patient.js]: Yeah, this is something I didn't think through enough. We definitely should _at least_ have a Result object type.

So maybe have an `Expected`, which a patient has one of for each population (and by default has one that has nothing expected to be passing), and also a `Results` object that a patient can have for each population. Ideally they could be indexed in an array on the patient by population number? Thoughts? Not sure how this might work in situations like on Dogpark, where patients are calculated against every measure (I don't quite recall how things work there).

I'm not sure about cached results actually.

J.B. was going to set up a meeting for Bonnie folks where I can go over this PR in detail. Perhaps we can add this discussion, as there are a few ways we can do this (and I'd like to know what people think makes sense for them).

Comment (jbradl11) [reply app/assets/javascripts/Patient.js]: Want to go over it after burn down? I didn't want to add another meeting to people's calendars.

Comment (holmesie) [reply app/assets/javascripts/Patient.js]: That works, but it should be a meeting for the folks that need to approve it. Everyone else can be optional.

File: templates/patient_extension.rb.erb [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+  def social_history
+    get_history_elements('social_history')
+  end
+
+  def vital_signs
+    get_history_elements('vital_sign')
+  end
+
+  def devices
+    get_history_elements('device')
```
Comment (mayerm94) [inline templates/patient_extension.rb.erb]: These should be re-sorted into alphabetical order by category type.  Looks like you are missing the 'symptom' type here.

File: app/models/qdm/patient_characteristic_sex.rb [hunk 17]
```diff
@@ -0,0 +1,4 @@
+class QDM::PatientCharacteristicSex
+  include Mongoid::Document
+  field :qdm_version, type: String, default: '5.3'
```
Comment (mayerm94) [inline app/models/qdm/patient_characteristic_sex.rb]: Should this include "  field :category, type: String, default: 'patient_characteristic'"?

Comment (holmesie) [reply app/models/qdm/patient_characteristic_sex.rb]: Per J.B.'s point above, yes, all of these will be `patient_caracteristic`s.

File: templates/patient_extension.rb.erb (hunk 16 above)
Comment (holmesie) [reply templates/patient_extension.rb.erb]: Good suggestion. And yes indeed, will need to add `symptom`.

File: templates/patient_extension.rb.erb [hunk 18]
```diff
@@ -0,0 +1,137 @@
+  field :ethnicity, type: Code
```
Comment (mayerm94) [inline templates/patient_extension.rb.erb]: (Posting for visibility) What was the motivation behind templating the patient?

File: app/models/qdm/patient.rb [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+
+  # These are the "data criteria", or QDM datatype elements that exist on a
```
Comment (okeefm) [inline app/models/qdm/patient.rb]: Are these `data_criteria` analogs, or `entry` analogs?

Comment (holmesie) [reply app/models/qdm/patient.rb]: `entry` analogs, I think I used `data_criteria` term that is what is coming from the HQMF parser (the draggable elements in the Bonnie patient builder are called `data_criteria`).

File: templates/patient_extension.rb.erb (hunk 18 above)
Comment (holmesie) [reply templates/patient_extension.rb.erb]: QDM actually already has a patient datatype. So we build the corresponding Patient model from what is in the modelinfo file, and then use this extension to add the additional functionality we need for Bonnie and Cypress. Using an ERB allows us to 1: separate out this code in a clean and easy to modify manner, and 2: allows us to at any point in the future make this extension more dynamic if need be.

File: app/assets/javascripts/Patient.js [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+PatientSchema.methods.transfers = function transfers(params, callback) {
+  return this.get_history_elements({'category': 'transfer'});
+}
+
+PatientSchema.methods.vital_signs = function vital_signs(params, callback) {
+  return this.get_history_elements({'category': 'vital_sign'});
+}
+
```
Comment (jbradl11) [inline app/assets/javascripts/Patient.js]: If the plan is to use this Patient Model with the CQL execution_engine, then we will need to add a "find_records" function, similar to what we have in the cql_qdm_patientapi, in the "cqlpatient.js.coffee" file.

Comment (holmesie) [reply app/assets/javascripts/Patient.js]: Right, same with `codes` on datatypes.

File: app/assets/javascripts/Patient.js [hunk 21]
```diff
@@ -0,0 +1,162 @@
+var Schema = mongoose.Schema;
+var Code = Schema.Types.Code;
+var Interval = Schema.Types.Interval;
+var Quantity = Schema.Types.Quantity;
+
+var PatientSchema = new Schema({
```
Comment (edeyoung) [inline app/assets/javascripts/Patient.js]: checking hds vs this.

title
religious affiliation
effective time
languages
test id
bundle id
marital status/record number/record assigner

does it make sense to include any of these? I lean to no, other than maybe bundle id. Does cypress have thoughts on bundle id?

**Comment by jbradl11** [comment]
I don't see a "Patient Source" model, which is another required wrapper model for the execution engine. Will this be added in this PR or at a later time?

**Comment by holmesie** [comment]
That's something that probably should happen here: https://github.com/projectcypress/js-ecqm-engine

File: app/models/qdm/diagnosis.rb [hunk 22]
```diff
@@ -0,0 +1,11 @@
+class QDM::Diagnosis < QDM::Datatype
+  include Mongoid::Document
+  field :author_datetime, type: DateTime
+  field :prevalence_period, type: QDM::Interval
+  field :anatomical_location_site, type: QDM::Code
+  field :severity, type: QDM::Code
+  field :hqmf_oid, type: String, default: '2.16.840.1.113883.10.20.28.3.110'
+  field :qrda_oid, type: String, default: '2.16.840.1.113883.10.20.24.3.135'
```
Comment (dczulada) [inline app/models/qdm/diagnosis.rb]: Why do some of these models have a qrda_oid, and others do not?

File: app/assets/javascripts/Patient.js [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+  sex: Code,
+  given_names: [String],
+  family_name: String,
+  bundle_id: String,
+  // These are the "data criteria", or QDM datatype elements that exist on a
+  // patient.
+  data_elements: []
+});
+
+// Returns an array of history elements that exist on this patient, that
```
Comment (ljades) [inline app/assets/javascripts/Patient.js]: You changed the `history_elements` terminology to `data_elements` in the code itself--the comments should be changed now too.

File: templates/patient_extension.js.erb [hunk 24]
```diff
@@ -0,0 +1,152 @@
+  ethnicity: Code,
+  race: Code,
+  sex: Code,
+  given_names: [String],
+  family_name: String,
+  bundle_id: String,
+  // These are the "data criteria", or QDM datatype elements that exist on a
+  // patient.
+  data_elements: []
```
Comment (edeyoung) [inline templates/patient_extension.js.erb]: i thought this was getting renamed

File: app/models/qdm/diagnosis.rb (hunk 22 above)
Comment (holmesie) [reply app/models/qdm/diagnosis.rb]: The OID list is not complete for QRDA, the field is there more to show that we will now have that info on the models.

File: app/assets/javascripts/Patient.js (hunk 23 above)
Comment (holmesie) [reply app/assets/javascripts/Patient.js]: Good catch

File: templates/patient_extension.js.erb [hunk 25]
```diff
@@ -0,0 +1,152 @@
+  ethnicity: Code,
```
[the remark this answers is not in the document]
Comment (holmesie) [reply templates/patient_extension.js.erb]: Right, removing these now.

File: templates/patient_extension.js.erb (hunk 24 above)
Comment (holmesie) [reply templates/patient_extension.js.erb]: Yeah it did get renamed, there a few comments that reflect the old naming though, and I am addressing that right now.
